File: pkg-py/tests/test_remote_pdb.py
"""PDB remote mode: live existence checks via the RCSB PDB data API."""

import pytest

import biobouncer
import biobouncer._remote as remote
from biobouncer import RemoteError


@pytest.fixture(autouse=True)
def _isolate_cache(tmp_path, monkeypatch):
    monkeypatch.setenv("BIOBOUNCER_CACHE_DIR", str(tmp_path))


def _stub(status, body=None):
    def _http_get(url, timeout=30):
        return status, body

    return _http_get


def test_url_builds_the_rcsb_entry_endpoint():
    from biobouncer._registry import get_source

    url = remote._pdb_url(get_source("pdb"), "4HHB")
    assert url == "https://data.rcsb.org/rest/v1/core/entry/4HHB"


def test_existing_structure_is_valid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(200))
    res = biobouncer.check_id("4HHB", source_db="pdb", how="remote")[0]
    assert res.valid is True
    assert res.normalized == "4HHB"
    assert res.suggestion is None


def test_absent_structure_is_invalid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(404))
    res = biobouncer.check_id("2ZZZ", source_db="pdb", how="remote")[0]
    assert res.valid is False
    assert res.suggestion is None


def test_lowercase_suggests_the_uppercase_form(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(200))
    res = biobouncer.check_id("4hhb", source_db="pdb", how="remote")[0]
    assert res.valid is False
    assert res.suggestion == "4HHB"


def test_malformed_skips_the_network(monkeypatch):
    def _forbidden(url, timeout=30):
        raise AssertionError("a malformed id must not reach the network")

    monkeypatch.setattr(remote, "_http_get", _forbidden)
    res = biobouncer.check_id("1ABCD", source_db="pdb", how="remote")[0]
    assert res.valid is False
    assert res.suggestion is None


def test_unexpected_status_raises_and_is_not_cached(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(500))
    with pytest.raises(RemoteError):
        biobouncer.check_id("4HHB", source_db="pdb", how="remote")
    assert not remote._remote_cache_path("pdb", "entry", "4HHB").is_file()

File: pkg-r/tests/testthat/test-remote-pdb.R
# Offline tests for pdb remote mode. Existence is checked against the RCSB PDB
# data API; the biobouncer.remote_transport option replaces it.

.stub_pdb <- function(status, body = NULL) {
  function(url, timeout) list(status = status, body = body)
}

test_that("the url addresses the rcsb entry endpoint", {
  resolver <- .remote_resolvers$pdb
  url <- resolver$url(NULL, "4HHB")
  expect_identical(url, "https://data.rcsb.org/rest/v1/core/entry/4HHB")
})

test_that("an existing structure is valid", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_pdb(200))
  res <- check_id("4HHB", "pdb", how = "remote")
  expect_true(res$valid)
  expect_identical(res$normalized, "4HHB")
  expect_true(is.na(res$suggestion))
})

test_that("an absent structure is invalid", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_pdb(404))
  res <- check_id("2ZZZ", "pdb", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("a lowercase id suggests the uppercase form", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_pdb(200))
  res <- check_id("4hhb", "pdb", how = "remote")
  expect_false(res$valid)
  expect_identical(res$suggestion, "4HHB")
})

test_that("a malformed id skips the network", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biobouncer.remote_transport = function(url, timeout) {
      stop("a malformed id must not reach the network")
    }
  )
  res <- check_id("1ABCD", "pdb", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("an unexpected status raises and is not cached", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_pdb(500))
  expect_error(
    check_id("4HHB", "pdb", how = "remote"),
    class = "biobouncer_error_remote"
  )
  expect_false(file.exists(.remote_cache_path("pdb", "entry", "4HHB")))
})
